hal/src/rtl872x/usb_hal.cpp

```diff
@@ -220,16 +223,24 @@ int32_t HAL_USB_USART_Send_Data(HAL_USB_USART_Serial serial, uint8_t data) {
     if ((__get_PRIMASK() & 1) || (__get_BASEPRI() != 0)) {
         return -1;
     }
-    int32_t available = -1;
-    do {
-        available = HAL_USB_USART_Available_Data_For_Write(serial);
-        if (available > 0) {
-            break;
+    int32_t available = HAL_USB_USART_Available_Data_For_Write(serial);
+    if (available == 0) {
+        bool needToYield = false;
+        auto prio = uxTaskPriorityGet(nullptr);
+        if (prio >= rtl::RTL_USBD_ISR_PROCESSING_THREAD_PRIORITY) {
+            needToYield = true;
         }
-        // TODO: check thread priorities here and delay only conditionally?
-        HAL_Delay_Milliseconds(1);
-    } while (available < 1 && available != -1);
-    if (HAL_USB_USART_Is_Connected(serial) && available > 0) {
+        while (true) {
+            available = HAL_USB_USART_Available_Data_For_Write(serial);
+            if (available > 0 || available < 0) {
+                break;
+            }
+            if (needToYield) {
+                HAL_Delay_Milliseconds(1);
+            }
+        }
```

Comment:
Just to be clear, the logic is as follows

- If the CDC driver has no room to write data, we wait for it to have room here
- If the driver has *any* room to write (`available >0`) or returns an error (`available < 0`) then break out of the loop and write any available data
- If the driver does not have room to write, and we need to yield (ie the thread calling this write function has higher priority than the USB ISR thread), then we yield so the ISR has a chance to send data and make room to write. 
- if the driver does not have room to write, and we are lower priority than the ISR thread, then we busy wait. The ISR task should preempt us and make room to write eventually.

Reply:
> If the driver has any room to write (available >0) or returns an error (available < 0) then break out of the loop and write any available data
Write only if `available > 0`, otherwise we simply error out.

Other things are correct.

Reply:
Is there any concern now that we are busy waiting? I guess to the user perspective its still blocking (busy loop vs delay previously). This only should affect the system if a user has a high priority thread doing this write (and/or they are constantly sending a lot of USB data)?

Reply:
Normally the TX buffer should clear up pretty quickly, so I wouldn't be too concerned and it's the exact same behavior as on other platforms since Gen 2 era. There are of course corner cases with host stopping to query for data ungracefully, in which case we should still get out of waiting state, but that might take a few seconds based on some internal timers we set up.